src/rgw/rgw_cache.cc

```diff
@@ -322,3 +337,453 @@ void ObjectCache::chain_cache(RGWChainedCache *cache) {
+
+  /*update cahce_map entries for new chunk in cache*/
+  cache_lock.Lock();
+  chunk_info->oid = oid;
+  chunk_info->set_ctx(cct);
+  chunk_info->size = len;
+  cache_map.insert(pair<string, ChunkDataInfo*>(oid, chunk_info));
+  cache_lock.Unlock();
+
+  return r;
+}
+
+void _cache_aio_write_completion_cb(sigval_t sigval) {
+
+	cacheAioWriteRequest *c = (cacheAioWriteRequest *)sigval.sival_ptr;
+	c->priv_data->cache_aio_write_completion_cb(c);
+}
+
+
+void DataCache::cache_aio_write_completion_cb(cacheAioWriteRequest *c){
+
+	ChunkDataInfo  *chunk_info = NULL;
+
+	ldout(cct, 0) << "engage: cache_aio_write_completion_cb oid:" << c->oid <<dendl;
+
+	/*update cahce_map entries for new chunk in cache*/
+	cache_lock.Lock();
+	outstanding_write_list.remove(c->oid);
+	chunk_info = new ChunkDataInfo;
+	chunk_info->oid = c->oid;
+	chunk_info->set_ctx(cct);
+	chunk_info->size = c->cb->aio_nbytes;
+	cache_map.insert(pair<string, ChunkDataInfo*>(c->oid, chunk_info));
+	cache_lock.Unlock();
+
+	/*update free size*/
+	eviction_lock.Lock();
+	free_data_cache_size -= c->cb->aio_nbytes;
+	outstanding_write_size -=  c->cb->aio_nbytes;
+	lru_insert_head(chunk_info);
+	eviction_lock.Unlock();
+	c->release();
+}
+
+int DataCache::create_aio_write_request(bufferlist& bl, unsigned int len, std::string oid){
+
+	struct cacheAioWriteRequest *wr= new struct cacheAioWriteRequest(cct);
+	int r=0;
+	if (wr->create_io(bl, len, oid) < 0) {
+		ldout(cct, 0) << "engage: Error create_aio_write_request" << dendl;
+		goto done;
+	}
+	wr->cb->aio_sigevent.sigev_notify = SIGEV_THREAD;
+	wr->cb->aio_sigevent.sigev_notify_function = _cache_aio_write_completion_cb;
+	wr->cb->aio_sigevent.sigev_notify_attributes = NULL;
+	wr->cb->aio_sigevent.sigev_value.sival_ptr = (void*)wr;
+	wr->oid = oid;
+	wr->priv_data = this;
+
+	if((r= ::aio_write(wr->cb)) != 0) {
+		ldout(cct, 0) << "ERROR: aio_write "<< r << dendl;
+		goto error;
+	}
+	return 0;
+
+error:
+	wr->release();
+done:
+	return r;
+}
+
+void DataCache::put(bufferlist& bl, unsigned int len, std::string oid){
+
+	int r = 0;
+	long long freed_size = 0, _free_data_cache_size = 0, _outstanding_write_size = 0;
+
+//	if (hash_uri(oid).compare(HASH_LOCAL) != 0) {
+//		dout(0) << "Engage1: Warning: We are only L2, no write:" << oid << dendl;
+//		return;
+//	}
+
+	ldout(cct, 0) << "Engage1: We are in DataCache::put() and oid is: " << oid <<dendl;
+	cache_lock.Lock();
+	map<string, ChunkDataInfo *>::iterator iter = cache_map.find(oid);
+	if (iter != cache_map.end()) {
+		cache_lock.Unlock();
+		ldout(cct, 0) << "Engage1: Warning: data already cached, no rewirte" << dendl;
+		return;
+	}
+	std::list<std::string>::iterator it = std::find(outstanding_write_list.begin(), outstanding_write_list.end(),oid);
+	if (it != outstanding_write_list.end()) {
+		cache_lock.Unlock();
+		ldout(cct, 0) << "Engage1: Warning: data put in cache already issued, no rewrite" << dendl;
+		return;
+	}
+	outstanding_write_list.push_back(oid);
+	cache_lock.Unlock();
```

Comment:
I do suspect that using std::unique_lock (or lock_guard, but it looks like we need more control) would improve this